Decryption failing on certain Android devices

Calling `getItem` fails with a `javax.crypto.AEADBadTagException` error on certain Android devices.

It happens for me on my Huawei P30 with Android 9, but doesn't happen on my Pixel 4 with Android 11, Samsung S10 with Android 11, Samsung S20 with Android 10, Samsung S9 with Android 9.

This appears to be related to some devices failing to decrypt when the message is too large (https://androidforums.com/threads/aeadbadtagexception-caused-by-keystoreexception-signature-mac-verification-failed.1327080/).


```
javax.crypto.AEADBadTagException
    at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:517)
    at javax.crypto.Cipher.doFinal(Cipher.java:2055)
    at dev.mcodex.RNSensitiveInfo.RNSensitiveInfoModule.decrypt(RNSensitiveInfoModule.java:677)
    at dev.mcodex.RNSensitiveInfo.RNSensitiveInfoModule.getItem(RNSensitiveInfoModule.java:195)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:372)
    at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:151)
    at com.facebook.react.bridge.queue.NativeRunnable.run(Native Method)
    at android.os.Handler.handleCallback(Handler.java:907)
    at android.os.Handler.dispatchMessage(Handler.java:105)
    at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
    at android.os.Looper.loop(Looper.java:216)
    at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:226)
    at java.lang.Thread.run(Thread.java:784)
Caused by: android.security.KeyStoreException: Signature/MAC verification failed
    at android.security.KeyStore.getKeyStoreException(KeyStore.java:851)
    at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:224)
    at android.security.keystore.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal(AndroidKeyStoreAuthenticatedAESCipherSpi.java:373)
    at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:506)
```

--- a/android/src/main/java/dev/mcodex/RNSensitiveInfoModule.java
+++ b/android/src/main/java/dev/mcodex/RNSensitiveInfoModule.java
@@ -29,6 +29,10 @@ import com.facebook.react.bridge.WritableMap;
 import com.facebook.react.bridge.WritableNativeMap;
 import com.facebook.react.modules.core.DeviceEventManagerModule;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
 import java.math.BigInteger;
 import java.security.InvalidKeyException;
 import java.security.Key;
@@ -45,6 +49,7 @@ import java.util.concurrent.Executors;
 
 import javax.crypto.BadPaddingException;
 import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
 import javax.crypto.IllegalBlockSizeException;
 import javax.crypto.KeyGenerator;
 import javax.crypto.SecretKey;
@@ -651,8 +656,25 @@ public class RNSensitiveInfoModule extends ReactContextBaseJavaModule {
             c = Cipher.getInstance(RSA_ECB);
             c.init(Cipher.ENCRYPT_MODE, publicKey);
         }
-        byte[] encodedBytes = c.doFinal(bytes);
-        String encryptedBase64Encoded = Base64.encodeToString(encodedBytes, Base64.DEFAULT);
+
+        int cipherTextSize = 0;
+        ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+        DataOutputStream dataStream = new DataOutputStream(byteStream);
+        ByteArrayInputStream plaintextStream = new ByteArrayInputStream(bytes);
+        final int chunkSize = 4*1024;
+        byte[] buffer = new byte[chunkSize];
+        while (plaintextStream.available() > chunkSize) {
+            int readBytes = plaintextStream.read(buffer);
+            byte[] ciphertextChunk = c.update(buffer, 0, readBytes);
+            cipherTextSize += ciphertextChunk.length;
+            dataStream.write(ciphertextChunk);
+        }
+        int readBytes = plaintextStream.read(buffer);
+        byte[] ciphertextChunk= c.doFinal(buffer, 0, readBytes);
+        cipherTextSize += ciphertextChunk.length;
+        dataStream.write(ciphertextChunk);
+
+        String encryptedBase64Encoded = Base64.encodeToString(byteStream.toByteArray(), Base64.NO_WRAP);
         return encryptedBase64Encoded;
     }
 
@@ -674,7 +696,21 @@ public class RNSensitiveInfoModule extends ReactContextBaseJavaModule {
             c = Cipher.getInstance(RSA_ECB);
             c.init(Cipher.DECRYPT_MODE, privateKey);
         }
-        byte[] decodedBytes = c.doFinal(Base64.decode(encrypted, Base64.DEFAULT));
+
+        byte[] bytes = Base64.decode(encrypted, Base64.NO_WRAP);
+        ByteArrayInputStream byteStream = new ByteArrayInputStream(bytes);
+        DataInputStream dataStream = new DataInputStream(byteStream);
+
+        CipherInputStream cipherStream = new CipherInputStream(byteStream, c);
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+        byte[] buffer = new byte[1024];
+        int len;
+        while ((len = cipherStream.read(buffer)) != -1) {
+            outputStream.write(buffer, 0, len);
+        }
+
+        byte[] decodedBytes = outputStream.toByteArray();
         return new String(decodedBytes);
     }
 }

